### PR TITLE
New Qoqo Gate Guide

### PR DESCRIPTION
**Pull request by mlodi-hqs**

### CODE REVIEW

**Review by dberthault — APPROVED** [review]
Looks good, maybe it would be nice to add something in the readme to say that if people wan to add new gates to qoqo there is this guide. Otherwise people might never find it

File: add_new_gate.md [hunk 1]
```diff
@@ -0,0 +1,491 @@
+# GUIDE: HOW TO ADD A NEW GATE TO ROQOQO/QOQO
+
+Depending on its type, adding support for a new quantum gate to roqoqo/qoqo can be more or less straightforward. As a rule of thumb, looking at the already available implementations of similar gates should help in understanding what to do.
+
+All supported gates are currently defined in the `roqoqo/src/operations` module (for qoqo, `qoqo/src/operations`). The specific script to put a new definition in depends on the type of gate.
+
+The general idea is: a `pub struct` defines the name, qubit-related fields as well as parameters (if any). Then, in order to add any kind of property to the gate, `traits` are either derived or manually implemented for it.
+
+As an examples, we'll take the roqoqo and qoqo definition of `RotateX`.
```
Comment (kbarkhqs) [inline add_new_gate.md:9]:
```suggestion
As an example, we'll take the roqoqo and qoqo definition of `RotateX`.
```

File: add_new_gate.md [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+        roqoqo_derive::SupportedVersion,
+        roqoqo_derive::Operate,
+        roqoqo_derive::Substitute,
+        roqoqo_derive::OperateSingleQubit,
+        roqoqo_derive::Rotate,
+    )]
+    #[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+    #[cfg_attr(feature = "json_schema", derive(schemars::JsonSchema))]
+    ```
+    Traits represent "properties" one can add to a gate. The idea, after defining the gate's `struct`, is to add these already defined properties to the gate.
```
Comment (kbarkhqs) [inline add_new_gate.md:128]:
```suggestion
    Traits represent "properties" one can add to a gate. The idea, after defining the gate's `struct`, is to add these previously defined properties to the gate.
```

File: add_new_gate.md [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+
+    Later on the guide will explain the meaning of all of these.
```
Comment (kbarkhqs) [inline add_new_gate.md:130]:
```suggestion
    Later on, the guide will explain the meaning of all of these.
```

File: add_new_gate.md [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+    ```
+
+    In general, structs define the signature of the gate: the name, the fields related to referencing qubits as well as the gate's parameter(s) (if any). In order to make the derivation of `traits` work, some of those fields are reserved, specifically:
+    ```rust
+    /// Array of field names that are reserved for use with specific traits
+    const RESERVED_FIELDS: &[&str; 15] = &[
+        "qubit",
+        "control",
+        "control_0",
+        "control_1",
```
Comment (kbarkhqs) [inline add_new_gate.md:149]: Missing control_2

File: add_new_gate.md [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+        "beta_r",
+        "beta_i",
+        "name",
+        "mode",
+        "mode_0",
+        "mode_1",
+    ];
+    ```
+    ##### In `roqoqo-derive/src/lib/rs`. If necessary, more can be added.
+    In this example, the `qubit` one. As for another example, a new two-qubit gate that is characterized by a `target` and `control` qubits will have to have those two fields (always of `usize` type) and the derivation of traits will work as expected, without having to do it manually.
```
Comment (kbarkhqs) [inline add_new_gate.md:164]:
```suggestion
    In this example, we use the `qubit` one. As for another example, a new two-qubit gate that is characterized by a `target` and `control` qubits will have to have those two fields (always of `usize` type) and the derivation of traits will work as expected, without having to do it manually.
```

File: add_new_gate.md [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+    #[allow(non_upper_case_globals)]
+    const TAGS_RotateX: &[&str; 5] = &[
+        "Operation",
+        "GateOperation",
+        "SingleQubitGateOperation",
+        "Rotation",
+        "RotateX",
+    ];
+    ```
+    These are related to the traits. In the example, "Operation" indicates that is a gate that implements the `Operate` trait; "GateOperation" the `OperateGate` trait; "SingleQubitGateOperation" the `OperateSingleQubitGate` trait and "Rotation" the `Rotate` trait. The last has to always be the gate's name.
```
Comment (kbarkhqs) [inline add_new_gate.md:177]:
```suggestion
    These are related to the traits. In the example, "Operation" indicates that is a gate that implements the `Operate` trait; "GateOperation" the `OperateGate` trait; "SingleQubitGateOperation" the `OperateSingleQubitGate` trait and "Rotation" the `Rotate` trait. The last tag must always be the gate's name. 
```

File: add_new_gate.md [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+        /// # Returns
+        ///
+        /// * `global_phase` - The global phase φ of the single-qubit unitary.
+        fn global_phase(&self) -> CalculatorFloat {
+            CalculatorFloat::from(0.0)
+        }
+    }
+    ```
+
+    These are gate-specific properties that cannot be derived. In this case, `OperateGate` and `OperateSingleQubitGate` are related to the matrix representation of the gate, so something that has be to manually defined.
```
Comment (kbarkhqs) [inline add_new_gate.md:244]:
```suggestion
    These are gate-specific properties that cannot be derived. In this case, `OperateGate` and `OperateSingleQubitGate` are related to the matrix representation of the gate, and are therefore something that has be to manually defined.
```

**Review by kbarkhqs — APPROVED** [review]

**Pull request merged**